src/herder/HerderImpl.cpp

```diff
@@ -264,10 +266,29 @@ HerderImpl::valueExternalized(uint64 slotIndex, StellarValue const& value)
     }
     else
     {
+        // Remember old LCL in case it changes
+        auto const lcl = mLedgerManager.getLastClosedLedgerNum();
+
         // This may trigger getting back in sync (buffered ledgers
         // application)
         processExternalized(slotIndex, value);
 
+        // If LM closed some ledgers, those are considered "validated" now
+        auto const newLcl = mLedgerManager.getLastClosedLedgerNum();
```

Comment:
This is why in the issue I was saying that maybe we need to have a callback from LM -> herder to notify herder that we just applied a ledger, and try to reason about that fact given the various things we know:
 * we can pass to herder the `LedgerCloseData` that was used to close the ledger (regardless of where it's coming from)
 * in herder we can query SCP to see if it was a non validated slot, and we can also use `getCurrentLedgerSeq()` to know if we should emit the message, to avoid emitting too many messages like you did

I think it's a simpler change that what you did here and easier to reason about.